### DISCUSSION TITLE
Interactive plots (tap evenet, select)

### DISCUSSION THREAD

**Discussion by kripnerl** [zauberzeug/nicegui#1598] [Ideas / Feature Requests]
Hello, 

first of all, thank you for this fantastic project! 

I would like to discuss the feasibility of support for interactive plots, which is crucial in my applications. 

# Example of application: 

* Select event occurrence in the data.
* Select the region of data to perform analysis (in the chart, not by slider). 
* Multidimensional view of data. 
* Design shapes/waveforms to be used in the subsequent workflow. 

# Current status in NiceGUI: 

To partly mentioned applications, currently, NiceGUI support  [Chart](https://nicegui.io/documentation/chart#chart_with_draggable_points) and [Interactive Images](https://nicegui.io/documentation/interactive_image). NiceGUI [Slider](https://nicegui.io/documentation#slider) sadly now does not support "range" option like the one in [Streamlit](https://docs.streamlit.io/library/api-reference/widgets/st.slider). 

While "chart" is usable for simple waveform generation. It is not ideal to plot various data over the chart. On the other hand, interactive images can be used to interact with matplotlib figures. But one is bound with the static figure and has to handle the position of the matplotlib axes in the figure. 

# Status in other projects 

All main plotting libraries support interactivity to some extend: 

- [Matplotlib](https://matplotlib.org/stable/users/explain/interactive.html) - even though support of interactivity in mpl is feasible, it is not ideal. 
- [Plotly/Dash](https://dash.plotly.com/interactive-graphing) - Some of the best interactions I have seen so far with fast responses. 
- [Bokeh](https://docs.bokeh.org/en/latest/docs/user_guide/interaction/js_callbacks.html) - Bokeh support JS callbacks. However, Bokeh[ is not yet supported](https://github.com/zauberzeug/nicegui/discussions/359) by NiceGUI. Example of interactions in [Holoviewes](https://holoviews.org/user_guide/Custom_Interactivity.html)
- [Vega](https://vega.github.io/vega-lite/examples/#interactive) 

# Suggested approach

While I am a physicist, not a JS/Web developer,  I may be misleading. However, the most straightforward way of achieving stated goals would be to develop an interaction layer for Plotly. 

# Final thoughts 

Even though LLM is strongly driving the development of all application frameworks. Hand data interaction/exploration is still crucial among other applications and is still tackled only in by few projects. It would be great if these features were seen in NiceGUI as well. 

Thank you very much if you put this into your routmap at some point. 

All the best, Lukas

**Comment by natankeddem** [comment]
So from what you are saying `ui.chart` (Highcharts) is not usable for your use case due to the way it renders/interacts. Are `ui.plotly` or `ui.echart` usable with additional event handling? What events specifically do you need? There is an active PR to bring a specific event interaction to `ui.echart` currently(https://github.com/zauberzeug/nicegui/pull/1579). 

If you can point to the specific interactions you are trying to achieve it would allow a better understanding of use case and best implementation of this functionality.

**Comment by kripnerl** [comment]
Hello, I was not aware of `echart`. It seems to be very powerful.  I will follow the implementation. Now I realise that both `chart` and `echart` are capable of far more than I was aware.  I'm not sure about the learning curve, though.  And performance when used on data with a high number of points. The syntax is not ideal either. Tools like `plotly_express` simplify one's life greatly. 

What is still missing is "data selection". A good example of what I mean is shown here: [Dash example](https://dash.plotly.com/interactive-graphing#update-graphs-on-hover). 

Enhanced interaction with `ui.plotly` would be indeed very appreciated. 

Thank you for your reply. I will try to use chart/echart for my proof of concept.

**Comment by natankeddem** [comment]
Could you list out the events you would find useful in the Plotly library reference?
https://plotly.com/javascript/plotlyjs-events/

**Reply by kripnerl** [reply]
In order of importance: 

1.  `plotly_click`
2. `plotly_selected` and `plotly_deselect`
3.  ... I can imagine the usage of those for better app functionality: 
4. `plotly_hover` and `plotly_unhover`
5. `plotly_doubleclick`
6. `plotly_selecting`
7. `plotly_legenclick` and  `plotly_legenddoubleclick`
8. ... I do not have any idea how to use the rest. But they may be eventually needed by somebody.

**Comment by falkoschindler** [comment]
We just merged PR #2231 which introduces arbitrary Plotly events, which was the main focus of this discussion. Therefore I'll close this feature request as resolved. Let me know if I missed something.
